Wiki reorg: ETH 2.0 comparison (#3789)
w3f/polkadot-wiki: docs/learn/learn-comparison-ethereum-2.md

---
id: learn-comparisons-ethereum-2
title: Polkadot vs. Ethereum
sidebar_label: Ethereum
description: A high-level comparison between Polkadot and Ethereum.
keywords: [ethereum, Ethereum, proof of stake, sharding]
slug: ../learn-comparisons-ethereum-2
---

> Polkadot and Ethereum are both **sharded** protocols

Reviewer: ```suggestion
Polkadot is the first fully sharded production-grade blockchain protocol.
```

> ([although Ethereum is still in the process of becoming a sharded network](https://notes.ethereum.org/@vbuterin/proto_danksharding_faq#What-is-Danksharding)).

Reviewer: ```suggestion
The current protocol of Ethereum does not implement sharding and it is the main focus of [the next major upgrade](https://notes.ethereum.org/@vbuterin/proto_danksharding_faq#What-is-Danksharding).
```

Similar to Polkadot and its parachains, Ethereum has goals of being a multi-sharded network. When a
network is **sharded**, this implies that it is capable of executing multiple (and often many) state

> transitions in parallel in a scalable manner.

Reviewer: ```suggestion
transitions in parallel in a scalable manner. One key difference is that Polkadot parachains are heterogeneous shards while Ethereum will have homogeneous shards. In other words, each parachain can have its own state transition logic whereas on Ethereum, all the shards use the same state transition logic.
```

Both protocols are blockchains but serve fundamentally different roles in how they are utilized:

- Ethereum is a general-purpose blockchain that hosts the Ethereum Virtual Machine, an environment
  for executing smart contracts.

> - Polkadot is a sharded, multi-chain protocol that hosts multiple chains and provides a way for them

Reviewer: ```suggestion
- Polkadot is a heterogeneous sharded, multi-chain protocol that hosts multiple chains and provides a way for them
```

  to partake in a shared security model. Polkadot acts as a **meta-protocol** that allows for
  multiple protocols to coexist and work together.

> :::warning

Reviewer: ```suggestion
:::info
```

data will be verifiable for an amount of time before being pruned from the network. These blobs will
have to be held for an amount of time, implying a level of data availability that validators must
have. This approach will enable data availability at layer one and further enable layer two
protocols on Ethereum to flourish.

In contrast, {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a purely sharded
network. It prioritizes data availability as an integral part of the block validation process.
Parallelized interactions between parachains, the **shards** of the Polkadot network, also take
advantage of this factor. Whereas Ethereum primarily focuses on making large amounts of data
available for validation for a portion of time, Polkadot's parallelization factor allows
verification to happen on the protocol level without needing a layer two solution.

:::note

The term "shards" is used here, but know that shards and parachains mean the same thing! Each
parachain represents a shard in the Polkadot network.

:::

On {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, each shard hosts core logic.

> However, each shard (i.e., [parachain](learn-parachains.md)) has a unique state transition function

Reviewer: ```suggestion
As mentioned earlier, each shard (i.e., [parachain](learn-parachains.md)) has a unique state transition function
```

(sometimes called a **runtime**). Applications can exist either within a single shard or across
shards by composing logic to create [cross-consensus (XCM)](learn-xcm.md) interactions.

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} uses WebAssembly
([Wasm](./learn-wasm.md)) as a "meta-protocol". A shard's state transition function can be abstract
as long as the validators on {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can
execute it within a Wasm environment.

## Architectural Differences: Polkadot and Ethereum

As previously mentioned, Ethereum is a **general-purpose** virtual machine that can run sandboxed
programs are written in Solidity, whereas Polkadot is a meta-protocol for other parachains to
connect and interact with each other.

> Ethereum operates as a single chain, homogenous chain. Each Ethereum node is divided into two

Reviewer: ```suggestion
Ethereum operates as a single chain, homogeneous chain. Each Ethereum node is divided into two
```

layers: the **consensus** and **execution** layers. Each layer handles the block validation
information, peer discovery, and Proof-of-Stake of the Ethereum client.

> Polkadot's primary component is the **relay chain**, which is a homogenous **sharded** network. The

Reviewer: ```suggestion
Polkadot's primary component is the **relay chain**, which hosts heterogeneous **shards** called parachains. The
```

relay chain aggregates information from its shards, the parachains, where Polkadot validators agree
upon consensus and finality. In essence, one can look at Polkadot as a series of **runtimes**, which
are state transition functions used to describe parachains (shards), as well as Polkadot itself.
Like Ethereum, Polkadot clients abstract away many of their responsibilities into various components
built using Substrate.

### Forks, Upgrades, and Governance

Ethereum governance is done off-chain, where various stakeholders come to a consensus through some
medium other than the protocol itself. Upgrades on Ethereum will follow the standard hard-fork
procedure, requiring validators to upgrade their nodes to implement protocol changes.

> {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} uses on-chain

Reviewer: This section needs to be revisited after OpenGov goes live on Polkadot

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can enact chain upgrades and
successful proposals using the Wasm meta-protocol without a hard fork. Anything within the state
transition function, the transaction queue, or off-chain workers can be upgraded without forking the
chain.

### Consensus and Finalization

Ethereum and {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} use hybrid consensus
models where block production and finality have their protocols. The finality protocols - Casper FFG
for Ethereum and [GRANDPA](./learn-consensus.md#finality-gadget-grandpa) for
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} - are both GHOST-based and can both
finalize batches of blocks in one round. For block production, both protocols use slot-based
protocols that randomly assign validators to a slot and provide a fork choice rule for unfinalized
blocks - RandDAO/LMD for Ethereum and [BABE](./learn-consensus.md#badass-babe-sassafras) for
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}.

There are two main differences between Ethereum and
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} consensus:

> 1.  Ethereum finalizes batches of blocks according to periods called "epochs ."The current plan is

Reviewer: ```suggestion
1.  Ethereum finalizes batches of blocks according to periods called "epochs". The current plan is
```

contracts can call each other; however, they are fixed on-chain to the domain of Ethereum. On
Polkadot, smart contracts can call each other in the same parachain and across parachains.

On Polkadot, developers have the option of either using smart contracts, calling extrinsics from
pallets that modify the chain's state in some particular way or merely use Polkadot's RPC to
directly retrieve and act on-chain information. DApps on Polkadot are often composed of these
multiple components working together to modify, retrieve, and watch state changes live as they
happen.

## Conclusion

Ethereum and {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} both use a sharded
model. Danksharding plans to utilize a rollup-centric approach by focusing on data availability. The
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} ecosystem is secured by a main chain,
called the "relay chain," which in turn manages and connects its shards ("parachains/parathreads")
into a single, homogenous solution.

The primary differences between the two protocols are:

- All shards in Ethereum represent the same state transitions, while

>   {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} lets shards have an abstract state

Reviewer: ```suggestion
  {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} lets shards be different and abstract away the state
```

>   transition function.

Reviewer: ```suggestion
  transition function implementation.
```

---
id: learn-comparisons-ethereum-2
title: Polkadot vs. Ethereum 2.0
sidebar_label: Ethereum 2.0
description: A high-level comparison between Polkadot and Ethereum 2.0.
keywords: [ethereum, ethereum 2.0, proof of stake, sharding]
slug: ../learn-comparisons-ethereum-2
---

Polkadot and Ethereum 2.0 are sharded blockchains as they provide both a) scalability by executing
transactions in separate shards and b) a protocol to send messages between shards.

## Model

The shards in Ethereum 2.0 have the same state transition function (STF), as the rules governing how
the blockchain can change state with each block are the same for all shards. Essentially, all shards
are identical copies. The STF provides an interface for smart contract execution, where contracts
exist on a single shard and can send asynchronous messages between shards.

> Likewise, on {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, each shard hosts core

Reviewer 2: Well, not _quite_ likewise, as not all STFs will be the same in theory :)

> logic, the shards are executed in parallel, and can send cross-shard asynchronous messages. However,

Reviewer 2: > can send cross-shard asynchronous messages

Is this supposed to relate to XCM?  If so, it'd be appropriate to link against XCM docs and clarify that.

each shard (i.e. [parachain](learn-parachains.md)) has a unique STF. Applications can exist either
within a single shard or across shards by composing logic.
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} uses WebAssembly
([Wasm](./learn-wasm.md)) as a "meta-protocol". A shard's STF can be abstract as long as the
validators on {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can execute it within a
Wasm environment. {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} supports smart
contracts through parachains. To offer some perspective, on Ethereum, smart contracts can call each
other synchronously in the same shard and asynchronously between shards. On Polkadot, smart

> contracts can call each other synchronously in the same parachain and asynchronously across

Reviewer 2: It might be beneficial to reference, or define what exactly *synchronous* and *asynchronous* could mean in this context, unless it is too much unneeded detail?

parachains.

## Architecture

### Ethereum 2.0

Ethereum 2.0's main chain is called the Beacon Chain. The primary load on the Beacon Chain is
attestations, which are votes on the availability of shard data and Beacon Chain validity. Each
shard is a blockchain with the Ethereum Wasm (eWasm) interface.

The initial plan prioritized the rollout of shards. Ethereum 2.0 launched Phase 0 of a multi-phase
rollout in December 2020, operating in parallel to the legacy Ethereum 1.0 chain:

- **Phase 0** provisioned the Beacon Chain, accepting deposits from validators and implementing
  proof-of-stake consensus, eventually among many shards.
- **Phase 1** launches 64 shards as simple chains to test the Beacon Chain's finality. Each shard
  submits "crosslinks" to the Beacon Chain, which contains the information to finalize shard data.
- **Phase 1.5** integrates Ethereum 1.0 as a shard to finalize the proof-of-work chain's blocks.
- **Phase 2** implements the eWasm interface, phasing out proof-of-work, finally making the system
  usable to end-users. See

>   [Ethereum 2.0 Phases](https://docs.ethhub.io/ethereum-roadmap/ethereum-2.0/eth-2.0-phases/) for

Reviewer 2: > [Ethereum 2.0 Phases](https://docs.ethhub.io/ethereum-roadmap/ethereum-2.0/eth-2.0-phases/) for

Bad link.

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} consensus:

1. Ethereum 2.0 finalizes batches of blocks according to periods called "epochs". The current plan
   is to have 32 blocks per epoch and finalize them all in one round. With a predicted block time of
   12 seconds, the expected time to finality is 6 minutes (12 minutes maximum). See
   [Ethereum 2 Block Time](https://github.com/ethereum/eth2.0-specs/blob/676e216/specs/phase0/beacon-chain.md#time-parameters)
   for more information.

   {{ polkadot: Polkadot's :polkadot }}{{ kusama: Kusama's :kusama }} finality protocol, GRANDPA,
   finalizes batches of blocks based on [availability and validity checks](./learn-availability.md)
   that happen as the proposed chain grows. The time to finality varies with the number of checks
   that need to be performed (and invalidity reports cause the protocol to require extra checks).
   The expected time to finality is 12-60 seconds.

2. Ethereum 2.0 requires many validators per shard to provide strong validity guarantees while
   {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can provide stronger guarantees

>    with fewer validators per shard. {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}

Reviewer 2: It would be nice to back this up with some data / number that is verifiable.

   achieves this by making validators distribute an
   [erasure coding](./learn-availability.md#erasure-codes) to all validators in the system, such
   that anyone - not only the shard's validators - can reconstruct a parachain's block and test its
   validity. The random parachain-validator assignments and secondary checks performed by randomly
   selected validators make it less likely for the small set of validators on each parachain to
   collude.

## Staking Mechanics

Ethereum 2.0 is a proof-of-stake network that requires 32 ETH to stake for each validator instance.
Validators run a primary Beacon Chain node and multiple validator clients - one for each 32 ETH.

> These validators get assigned to "committees", randomly selected groups to validate shards in the

Reviewer 2: Link to what a "committee" in Ethereum is and make it more explicit.

network. Ethereum 2.0 relies on having a large validator set to provide availability and validity
guarantees: They need at least 111 validators per shard to run the network and 256 validators per
shard to finalize all shards within one epoch. With 64 shards, that's 16_384 validators (given 256
validators per shard). See
[Ethereum 2.0 Economics](https://docs.ethhub.io/ethereum-roadmap/ethereum-2.0/eth-2.0-economics/)
and [Eth2 shard chain simplification proposal](https://notes.ethereum.org/@vbuterin/HkiULaluS) for
more information.

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can provide strong finality and
availability guarantees with much fewer validators. It uses
[Nominated Proof of Stake (NPoS)](learn-staking.md) to select validators from a smaller set, letting
smaller holders nominate validators to run infrastructure while still claiming the rewards of the
system without running a node of their own.
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} needs about ten validators for each
parachain in the network.

## Shards

> Every shard in Ethereum 2.0 has the same STF. Each shard will submit "crosslinks" to the beacon

Reviewer 2: Link to what a "crosslink" is, or some other resource.

chain and implement an eWasm execution environment. EWasm is a restricted subset of Wasm for
contracts in Ethereum. The eWasm interface provides a set of methods available to contracts. There
should be a similar set of development tools like Truffle and Ganache to develop for eWasm. See
[eWasm Design](https://github.com/ewasm/design) for more information. See
[Sharding FAQ](https://github.com/ethereum/wiki/wiki/Sharding-FAQ#how-would-synchronous-cross-shard-messages-work)
for more information.

Every shard in {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} has an abstract STF
based on Wasm. Each shard can expose a custom interface as long as the logic compiles to Wasm and
the shard provides an "execute block" function to
{{ polkadot: Polkadot's :polkadot }}{{ kusama: Kusama's :kusama }} validators. Polkadot’s
[Substrate](https://substrate.io/) development framework allows full spectrum composability with a

> suite of modules that can be configured, composed, and extended to develop a chain's STF.

Reviewer 2: Mention how this is also called a "runtime" here? They are synonymous, but maybe not obvious to a first time reader.

> ## Rollups

Reviewer 2: Unclear transition to how rollups relate to the previous content. IMO should be separated into its own sub-category under the sidebar.

bitcoin up on-chain, then puts it in channels, that are all shared between different nodes. Bitcoin
can then be routed across these channels. There's no "lightning blockchain".

### Optimistic Rollups

Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction
commitments to the network while validating for errors and security issues. If something is invalid,
validators can submit fraudulent commitment proof (fraud-proofs). These proofs will be used to
reject fraudulent transactions. This way, transaction results and proofs can be stored on the main
network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction
scalability 10-100x and write transactions to Ethereum as
[`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata), reducing gas fees
significantly. Examples of optimistic rollup solutions include
[Optimisim](https://www.optimism.io/), [Arbitrum](https://bridge.arbitrum.io/) and
[Unipig](https://unipig.exchange/welcome). See
[ETH Docs: Optimistic Rollups](https://ethereum.org/en/developers/docs/scaling/optimistic-rollups/)
for more information.

> ### Zero-knowledge Rollups

Reviewer 2: I would shorten this, add new info, and make it its own h2 section with what ETH 2.0 and Polkadot are doing underneath respectively.  Might be wise to include some of the parachains who are working with ZK proofs in here as well?

to set the passing threshold for low-turnout referenda. Referenda can cover various topics,
including fund allocation from an on-chain [Treasury](./learn-treasury.md) or modifying the
underlying runtime code of the chain. Decisions get enacted on-chain and are binding and autonomous.

## Upgrades

Upgrades on Ethereum 2.0 will follow the standard hard-fork procedure, requiring validators to
upgrade their nodes to implement protocol changes.

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} can enact chain upgrades and
successful proposals using the Wasm meta-protocol without a hard fork. Anything within the STF, the
transaction queue, or off-chain workers can be upgraded without forking the chain.

## Conclusion

Ethereum 2.0 and {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} both use a sharded
model where shard chains ("shards" in Ethereum 2.0 and parachains in
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}) are secured by a main chain by
linking shard state in the blocks of the main chains. Main differences between the two protocols
are:

- All shards in Ethereum 2.0 have the same STF, while

>   {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} lets shards have an abstract STF.

Reviewer 2: Emphasize that the allowance of an abstract STF would allow for more specialized chains to come into play, versus a singular, boilerplate STF.

validators. Polkadot has the Substrate development framework that allows full spectrum composability
with a suite of modules that can be configured, composed, and extended to develop a chain's STF.

## Rollups

> Rollups, a layer 2 scaling solution, have been talked about for some time and it's potential to scaling Ethereum. A layer 2 network of nodes would be responsible for "rolling up" transaction together by batching them before publishing them to the main Ethereum chain. A rolled up transaction could inclue 10,000 transactions. Rollups increase TPS and can reduce transaction fees significantly. When it comes to implementing it, there are a few different ways to approach the problem.

Reviewer: ```suggestion
Rollups, a layer 2 scaling solution, have been talked about for some time, and their potential to scale Ethereum. A layer 2 network of nodes would be responsible for "rolling up" transaction together by batching them before publishing them to the main Ethereum chain. A rolled-up transaction could include 10,000 transactions. Rollups increase TPS and can reduce transaction fees significantly. When it comes to implementing it, there are a few different ways to approach the problem.
```

### Optimistic Rollups

> Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction commitments to the network while validating for errors and security issues. If invalid, validators can submit a fradulent commitment proof(fraud-proofs). These proofs will be used to reject fradulant transactions. This way, transaction results along with a proof can be stored on the main network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction scalability 10-100x, and writes transactions to Ethereum as [`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata) reducing gas fees significantly.

Reviewer: ```suggestion
Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction commitments to the network while validating for errors and security issues. If invalid, validators can submit a fraudulent commitment proof (fraud-proofs). These proofs will be used to reject fraudulent transactions. This way, transaction results along with a proof can be stored on the main network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction scalability 10-100x, and write transactions to Ethereum as [`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata), reducing gas fees significantly.
```

> Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction commitments to the network while validating for errors and security issues. If invalid, validators can submit fraudulent commitment proof (fraud-proofs). These proofs will be used to reject fraudulent transactions. This way, transaction results along with proof can be stored on the main network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction scalability 10-100x, and write transactions to Ethereum as [`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata), reducing gas fees significantly.

Reviewer 3: 'If invalid' sentence grammar is off.

'10-100x' needs a source.

Author: Source is ETH docs https://ethereum.org/en/developers/docs/scaling/optimistic-rollups/ I've added it to the resource section

Examples of optimistic rollup solutions include [Optimisim](https://www.optimism.io/), [Arbitrum](https://bridge.arbitrum.io/) and [Unipig](https://unipig.exchange/welcome).

### Zero-knowledge Rollups

> [Zero-knowledge proofs](https://en.wikipedia.org/wiki/Zero-knowledge_proof) can be used as a "proof" of valid transactions when batching, effectively reducing the transaction payload to be the size of a zk-proof. This, similar to the optimistic approach, allows transaction data to be abstracted away and reduced significantly in size before being committed to the main chain. Similar to most on-chain transactions and states, a ZK-rollup's state is represented as a [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree), and the cryptographic hash of the tree's root(Merkle root) is what is stored on-chain. This allows for efficient batching, and tracking of changes to the transaction states. Any changes to the state will require the operator who initiated the changes to compute a new state root and commit it to the on-chain contract. If the contract verifies the new state as valid, then provided root hash becomes the new state root.

Reviewer 3: It's worth mentioning the turing-completeness issue with ZK here.

Author: I looked for some good resources online about the completeness problem, but couldn't find a good piece. Let me know if you have any that we can add to the resources.

Reviewer 4: TLD;DR
Expressing arbitrary computation inside a circuit to be proved in zero knowledge is theoretically possible, but slow & cumbersome.

Layer-1:
To reach consensus on a traditional block execution, blockchains rely on re-execution of every transaction. We don't want to allow malicious transactions that cause an endless loop, as it would compromise every miner's/validator's resources and the liveness of the chain. Therefore chains like Ethereum have the concept of gas, so the amount of computation per block is bounded. So even though you might have a notion of Turing completeness, there's actually a hard bound on how many operations you can perform. In addition, as the number of miners/validators scales, the amount of redundant computation scales linearly too, which is bad.

Now on to ZK rollups: 
Good thing is that you avoid the last point, i.e. redundant execution of all transactions. The majority of the network only runs a `verify` operation that should be sublinear in the size of the original Tx's, while the block producer incurs heavy computation to produce the proof.
The challenge for any proof system architect is expressing the (arbitrary) computation as a circuit. For a number of reasons (e.g. all arithmetic needs to be done over a finite field), the circuits that the proof systems "accept" are fairly large and are slower to work with than executing the computation directly (aka natively), requiring up to 1,000,000x work for the prover.
If you want to execute an extremely complicated program, you might run into circuits which are simply too large for the prover to handle, or take too long.

A good article on the limitations of ZK proving systems: https://a16zcrypto.com/measuring-snark-performance-frontends-backends-and-the-future/

#### EIP 4844

"Introduce a new transaction format for “blob-carrying transactions” which contain a large amount of data that cannot be accessed by EVM execution, but whose commitment can be accessed. The format is intended to be fully compatible with the format that will be used in full sharding."

Until full spec of ETH 2.0 sharding is implemented, this stopgap introduces "blob-transactions" transaction format in the same way they would be designed with sharding, but without sharding those transactions. Blob-transactions are transactions which carry a large amount of data that cannot be accessed by the EVM, but whose commitment can be accessed. Compared to full sharding where the space allocated for this additional transaction data will be ~16mb, this EIP allows for a reduced cap corresponding to a target of ~1mb and a limit of ~2mb which the rollup transactions can utilize. 

### Rollups in Polkadot

> In Polkadot the sharding model already exists as the Relay-chain being the beacon chain  and Parachains being the shards. Parachains are most similar in implementation to an Optimistic rollup and most similar in architecture to a ZK-rollup. Parachain logic already run a validity proof. The proof (the approvals protocol) is interactive, unlike ZK-rollups which are non-interactive. Aditionally, unlike ZK-rollups, there are no difficulties in creating parachains with turing-complete logic. This is a fundamental weakness of ZK rollups, as turing completeness within ZK circuits is not easy.

Reviewer: ```suggestion
In Polkadot, the sharding model already exists as the Relay-chain being the beacon chain and Parachains being the shards. Parachains are most similar in implementation to an Optimistic rollup and in architecture to a ZK-rollup. Parachain logic already runs a validity proof. The proof (the approvals protocol) is interactive, unlike ZK-rollups, which are non-interactive. Additionally, unlike ZK-rollups, there are no difficulties in creating parachains with Turing-complete logic. This is a fundamental weakness of ZK rollups, as Turing completeness within ZK circuits is not easy.
```

> In Polkadot, the sharding model already exists as the Relay-chain being the beacon chain and Parachains being the shards. Parachains are most similar in implementation to an Optimistic rollup and in architecture to a ZK-rollup. Parachain logic already runs a validity proof. The proof (the approvals protocol) is interactive, unlike ZK-rollups, which are non-interactive. Additionally, unlike ZK-rollups, there are no difficulties in creating parachains with Turing-complete logic. This is a fundamental weakness of ZK rollups, as Turing completeness within ZK circuits is not easy.

Reviewer 3: 'Parachain logic already...' could be edited for clarity. style-wise I'd omit 'already'.

Would be nice to link here.

We should clarify in this section that Parachains are akin to rollups and could even be considered as a category of rollup.

Author: Not sure if there is good literature out there yet to link something here

### Optimistic Rollups

Using an "innocent until proven guilty" approach, optimistic rollups will accept transaction commitments to the network while validating for errors and security issues. If something is invalid, validators can submit fraudulent commitment proof (fraud-proofs). These proofs will be used to reject fraudulent transactions. This way, transaction results along with proof can be stored on the main network, but transaction data can be stored elsewhere. Optimistic rollups can improve transaction scalability 10-100x, and write transactions to Ethereum as [`calldata`](https://ethereum.stackexchange.com/questions/52989/what-is-calldata), reducing gas fees significantly.

Examples of optimistic rollup solutions include [Optimisim](https://www.optimism.io/), [Arbitrum](https://bridge.arbitrum.io/) and [Unipig](https://unipig.exchange/welcome).

### Zero-knowledge Rollups

:::note

> Currently, Turing completeness is difficult in ZK rollups, a trusted setup is required to run the ZK Proofs, as well as the soundness of the cryptography, used needs to meet certain standards.

Reviewer 4: Trusted setup is not necesssarily required, depending on which type of a system you use.
See [STARK vs SNARK](https://consensys.net/blog/blockchain-explained/zero-knowledge-proofs-starks-vs-snarks/).

> Currently, Turing completeness is difficult in ZK rollups, a trusted setup is required to run the ZK Proofs, as well as the soundness of the cryptography, used needs to meet certain standards.

Reviewer 4: Also the part about cryptography & standards doesn't make sense to me in its current form.

#### EIP 4844

"Introduce a new transaction format for “blob-carrying transactions” which contain a large amount of data that cannot be accessed by EVM execution, but whose commitment can be accessed. The format is intended to be fully compatible with the format that will be used in full sharding."

> Until the full spec of ETH 2.0 sharding is implemented, this stopgap introduces the "blob-transactions" transaction format, in the same way they would be designed with sharding, but without sharding those transactions. Blob-transactions are transactions that carry a large amount of data that cannot be accessed by the EVM, but whose commitment can be accessed. Compared to full sharding where the space allocated for this additional transaction data will be ~16MB, this EIP allows for a reduced cap corresponding to a target of ~1MB and a limit of ~2MB, which the rollup transactions can utilize. 

Reviewer 4: Some of the text is basically a repetition of the previous paragraph.

## Rollups

> Rollups, a layer 2 scaling solution, have been talked about for some time, and their potential to scale Ethereum. A layer 2 network of nodes would be responsible for "rolling up" transaction together by batching them before publishing them to the main Ethereum chain. A rolled-up transaction could include 10,000 transactions. Rollups increase TPS and can reduce transaction fees significantly. When it comes to implementing it, there are a few different ways to approach the problem.

Reviewer 3: 10,000 seems to be an arbitrary choice.

style: layer-2.

typo: "rolling up" transaction**s**

Either the transactions themselves are published _or_ a commitment to them. It's more typical for a commitment to be published to the chain and the transactions themselves to be published to a DA layer. In the case of ZK rollups, the transactions may not need to be published - only the changes to the rollup chain-state.

I would replace the last sentence with "Rollups are a broad category of solution which encompass many different possible architectures and implementations."

Author: 10k was from the ETH docs, but yes I think it's somewhat arbitrary and will replace it with something more flexible
